### PR TITLE
Issue 23159 allow add or remove contentlet to a specific pages version variant

### LINKED ISSUE #23159: Allow Add or Remove contentlet to a specific page's version variant
### Is your feature request related to a problem? Please describe.
When Add or remove a contentlet in a page it is Add or Remove for all the page's versions.

### Describe the solution you'd like
When Add or Remove a contentlet into a page's version, it should be add or remove just for the page's version that we are working on.

By now we need that this just work for page experiment variant version.

### Additional context
I think we can create two new fields: add_just, remove_just into multi_tree table, there we can have a list (string split by comma) of the inode where the contentlet is added or removed, maybe it can be just one field and use a JSON value to set both values.

### Acceptance criteria
- The contentlet should be Add or Remove just for the page experiement variant version.
- Integration test.
- Don't broke anything, we need to be sure that we don't broke anything in the render proccess

### PR DESCRIPTION
**Pull request by freddyucv**
### Proposed Changes
* Add a new variant_id field into the multi_tree table
* Also use this new field into MultiTreeAPIImpl
* Move some methods to DotDatabaseMetaData to reuse
* Modify the SLECT that we have right now to hit the multi_tree table, they were update to add the variant_id field in different ways:
- Add variant as parameters because it is use into the save variant API method to check if the primary key is not duplicated

https://github.com/dotCMS/core/pull/23211/files#diff-926e9e9e22e2c282fa305bbe50022e93d9d91d6b83a5c33c8e29cacde4a0b52cR82

- This one is used to delete all the multi_tree before add the new ones, when we change any multi_tree into the page really what we are doing is deleting all the multi_tree and then add all of them again

https://github.com/dotCMS/core/pull/23211/files#diff-926e9e9e22e2c282fa305bbe50022e93d9d91d6b83a5c33c8e29cacde4a0b52cR88

- For the others was added the filter to get just the DEFAULT versions, because:
Some of them are been using into API methods that we are not using into our code but the client can be using them so we don't want to change the behavior for this method right now.
Also some then are be using into features like Push Publish and even though we need to change the behavior of PP to send new concept like Experiment, Variants, etc, I think it is better to change it when we will be really working on that

 https://github.com/dotCMS/core/pull/23211/files#diff-926e9e9e22e2c282fa305bbe50022e93d9d91d6b83a5c33c8e29cacde4a0b52cL95-L104

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java [hunk 1]
```diff
@@ -699,6 +700,54 @@ private List<String> getIndicesMSSQL(final String tableName, final DotConnect do
         return indexMeta.stream().map(map -> map.get("index_name").toString()).collect(Collectors.toList());
     }
 
+    public static List<String> getPrimaryKeysFields(final String tableName)  {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java:703]: doc me

File: dotCMS/src/integration-test/java/com/dotmarketing/factories/MultiTreeAPITest.java [hunk 2]
```diff
@@ -143,16 +154,82 @@ public  void testDeletes() throws Exception {
...
         List<MultiTree> list = APILocator.getMultiTreeAPI().getMultiTrees(PAGE, CONTAINER+0, RELATION_TYPE+0);
+
+        assertTrue("multiTree reorders", list.get(0).getContentlet().equals("CONTENTLET1"));
+        assertTrue("multiTree reorders", list.get(1).getContentlet().equals("CONTENTLET2"));
         assertTrue("multiTree reorders", list.get(2).equals(tree));
+        assertTrue("multiTree reorders", list.get(3).getContentlet().equals("CONTENTLET3"));
+        assertTrue("multiTree reorders", list.get(4).getContentlet().equals("CONTENTLET4"));
+    }
+
+    @Test
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/integration-test/java/com/dotmarketing/factories/MultiTreeAPITest.java:175]: doc test scenario and expected results

File: dotCMS/src/main/java/com/dotmarketing/beans/MultiTree.java [hunk 3]
```diff
@@ -65,7 +70,18 @@ public MultiTree(final String htmlPage,
                      final String instanceId,
                      final int treeOrder) {
 
-        this(htmlPage, container, child, instanceId, treeOrder, DOT_PERSONALIZATION_DEFAULT);
+        this(htmlPage, container, child, instanceId, treeOrder, DOT_PERSONALIZATION_DEFAULT,
+                VariantAPI.DEFAULT_VARIANT.name());
+    }
+
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotmarketing/beans/MultiTree.java:76]: better doc here. Like what every param does etc.

File: dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java [hunk 4]
```diff
@@ -699,6 +700,54 @@ private List<String> getIndicesMSSQL(final String tableName, final DotConnect do
...
+                            + "AND constraint_type = 'PRIMARY KEY'")
+                    .addParam(tableName)
+                    .loadResults();
+
+            return arrayList.isEmpty() ? Optional.empty() :
+                    Optional.of((((Map) arrayList.get(0))).get("constraint_name").toString());
+        } catch (DotDataException e) {
+            return Optional.empty();
+        }
+    }
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java:750]: if you moved them here you should probably add test to DotDatabaseMetaDataTest too

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java [hunk 5]
```diff
@@ -552,7 +561,8 @@ public void overridesMultitreesByPersonalization(final String pageId,
     public void overridesMultitreesByPersonalization(final String pageId,
                                                     final String personalization,
                                                     final List<MultiTree> multiTrees,
-                                                     final Optional<Long> languageIdOpt) throws DotDataException {
+                                                    final Optional<Long> languageIdOpt,
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java:564]: Optional as params is not a good practice

**Review by fabrizzio-dotCMS — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java [hunk 6]
```diff
@@ -620,15 +634,24 @@ public void overridesMultitreesByPersonalization(final String pageId,
...
+            Optional<Long> languageIdOpt) throws DotDataException {
+        overridesMultitreesByPersonalization(pageId, personalization, multiTrees,
+                languageIdOpt, VariantAPI.DEFAULT_VARIANT.name());
+    }
+
     private void deleteMultiTreeToMySQL(
             final String pageId,
             final String personalization,
-            final Optional<Long> languageIdOpt) throws DotDataException {
+            final Optional<Long> languageIdOpt, String variantId) throws DotDataException {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java:648]: final

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java (hunk 5 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java:564]: really it was like this before my changes, and to change I have to change it in several places

File: dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotmarketing/common/db/DotDatabaseMetaData.java:709]: done https://github.com/dotCMS/core/pull/23211/files#diff-9516bc9da076af076365fb4fdd4816c46a9d733523c29f4e7ff1ff7c69a974ebR92

**Review by jdotcms — APPROVED** [review]

**Review by fabrizzio-dotCMS — APPROVED** [review]

**Pull request merged**